File: GanServer/src/com/xattacker/binary/InputBinaryBuffer.java
package com.xattacker.binary;

import java.io.Closeable;
import java.io.IOException;
import java.io.InputStream;

public class InputBinaryBuffer implements Closeable, BinaryReadable
{
	private InputStream _in;

	public InputBinaryBuffer(InputStream aIn)
	{
		_in = aIn;
	}
	
	public long available()
	{
		int available = 0;
		
		try
		{
			available = _in.available();
		}
		catch (Exception ex)
		{
			available = 0;
		}

		return available;
	}

	public void close() throws IOException
	{
		if (_in != null)
		{
			_in.close();
		}
	}

	public int read(byte[] aData) throws IOException
	{
		return _in.read(aData);
	}
	
	public int read(byte[] aData, int aOffset, int aLength) throws IOException
	{
		return _in.read(aData, aOffset, aLength);
	}

	public byte[] readBinary(int aLength) throws IOException
	{
		if (_in.available() < aLength)
		{
			throw new IOException("Failed to read binary data from a input stream (Index out of bound)");
		}

		byte[] data = new byte[aLength];
		_in.read(data);

		return data;
	}

	public short readShort() throws IOException
	{
		if (_in.available() < 2)
		{
			throw new IOException("Failed to read a short from a input stream (Index out of bound)");
		}
		else
		{
			byte[] data = new byte[2];
			_in.read(data);

			return TypeConverter.byteToShort(data);
		}
	}

	public int readInteger() throws IOException
	{
		if (_in.available() < 4)
		{
			throw new IOException("Failed to read a integer from a input stream (Index out of bound)");
		}
		else
		{
			byte[] data = new byte[4];
			_in.read(data);

			return TypeConverter.byteToInt(data);
		}
	}

	public long readLong() throws IOException
	{
		if (_in.available() < 8)
		{
			throw new IOException("Failed to read a long from a input stream (Index out of bound)");
		}
		else
		{
			byte[] data = new byte[8];
			_in.read(data);

			return TypeConverter.byteToLong(data);
		}
	}

	public double readDouble() throws IOException
	{
		if (_in.available() < 8)
		{
			throw new IOException("Failed to read a double from a input stream (Index out of bound)");
		}
		else
		{
			byte[] data = new byte[8];
			_in.read(data);

			return TypeConverter.byteToDouble(data);
		}
	}

	public String readString() throws IOException
	{
		int length = readInteger();

		if (length == 0)
		{
			return "";
		}

		byte[] data = readBinary(length);

		return new String(data, "UTF8");
	}
}

File: GanAndroidApp/app/src/main/java/com/xattacker/binary/InputBinaryBuffer.kt
package com.xattacker.binary

import java.io.Closeable
import java.io.IOException
import java.io.InputStream

class InputBinaryBuffer: Closeable, BinaryReadable
{
    private var _in: InputStream?

    constructor(ins: InputStream)
    {
        _in = ins
    }

    override fun available(): Long
    {
        var available = 0

        try
        {
            available = _in?.available() ?: 0
        }
        catch (ex: Exception)
        {
            available = 0
        }

        return available.toLong()
    }

    @Throws(IOException::class)
    override fun close()
    {
        _in?.close()
    }

    @Throws(IOException::class)
    fun read(aData: ByteArray): Int?
    {
        return _in?.read(aData)
    }

    @Throws(IOException::class)
    fun read(aData: ByteArray, aOffset: Int, aLength: Int): Int?
    {
        return _in?.read(aData, aOffset, aLength)
    }

    @Throws(IOException::class)
    override fun readBinary(aLength: Int): ByteArray?
    {
        if (_in?.available() ?: 0 < aLength)
        {
            throw IOException("Failed to read binary data from a input stream (Index out of bound)")
        }

        val data = ByteArray(aLength)
        _in?.read(data)

        return data
    }

    @Throws(IOException::class)
    override fun readShort(): Short?
    {
        if (_in?.available() ?: 0 < TypeConverter.SHORT_SIZE)
        {
            throw IOException("Failed to read a short from a input stream (Index out of bound)")
        }
        else
        {
            val data = ByteArray(TypeConverter.SHORT_SIZE)
            _in?.read(data)

            return TypeConverter.byteToShort(data)
        }
    }

    @Throws(IOException::class)
    override fun readInteger(): Int?
    {
        if (_in?.available() ?: 0 < TypeConverter.INTEGER_SIZE)
        {
            throw IOException("Failed to read a integer from a input stream (Index out of bound)")
        }
        else
        {
            val data = ByteArray(TypeConverter.INTEGER_SIZE)
            _in?.read(data)

            return TypeConverter.byteToInt(data)
        }
    }

    @Throws(IOException::class)
    override fun readLong(): Long?
    {
        if (_in?.available() ?: 0 < TypeConverter.LONG_SIZE)
        {
            throw IOException("Failed to read a long from a input stream (Index out of bound)")
        }
        else
        {
            val data = ByteArray(TypeConverter.LONG_SIZE)
            _in?.read(data)

            return TypeConverter.byteToLong(data)
        }
    }

    @Throws(IOException::class)
    override fun readDouble(): Double?
    {
        if (_in?.available() ?: 0 < TypeConverter.DOUBLE_SIZE)
        {
            throw IOException("Failed to read a double from a input stream (Index out of bound)")
        }
        else
        {
            val data = ByteArray(TypeConverter.DOUBLE_SIZE)
            _in?.read(data)

            return TypeConverter.byteToDouble(data)
        }
    }

    @Throws(IOException::class)
    override fun readString(): String?
    {
        var str: String? = null

        val length = readInteger()
        if (length != null)
        {
            val data = readBinary(length)
            if (data != null)
            {
                str = String(data, Charsets.UTF_8)
            }
        }

        return str
    }
}
